lib/Dialect/FSM/FSMOps.cpp

```diff
@@ -0,0 +1,373 @@
+  if (referencedMachine == nullptr)
+    return op.emitError("Cannot find machine definition '")
+           << op.machine() << "'";
+
+  // Check operand types first.
+  auto numOperands = op->getNumOperands();
+  auto expectedOperandTypes = referencedMachine.getType().getInputs();
+
+  if (expectedOperandTypes.size() != numOperands) {
+    auto diag = op.emitOpError()
+                << "has a wrong number of operands; expected "
+                << expectedOperandTypes.size() << " but got " << numOperands;
+    diag.attachNote(referencedMachine->getLoc())
+        << "original machine declared here";
+
+    return failure();
+  }
+
+  for (size_t i = 0; i != numOperands; ++i) {
+    auto expectedType = expectedOperandTypes[i];
+    auto operandType = op.getOperand(i).getType();
+    if (operandType != expectedType) {
+      auto diag = op.emitOpError()
+                  << "#" << i << " operand type must be " << expectedType
+                  << ", but got " << operandType;
+
+      diag.attachNote(referencedMachine->getLoc())
+          << "original machine declared here";
+      return failure();
+    }
+  }
+
+  // Check result types.
+  auto numResults = op->getNumResults();
+  auto expectedResultTypes = referencedMachine.getType().getResults();
+
+  if (expectedResultTypes.size() != numResults) {
+    auto diag = op.emitOpError()
+                << "has a wrong number of results; expected "
+                << expectedResultTypes.size() << " but got " << numResults;
+    diag.attachNote(referencedMachine->getLoc())
+        << "original machine declared here";
+
+    return failure();
+  }
+
+  for (size_t i = 0; i != numResults; ++i) {
+    auto expectedType = expectedResultTypes[i];
+    auto resultType = op.getResult(i).getType();
+    if (resultType != expectedType) {
+      auto diag = op.emitOpError()
+                  << "#" << i << " result type must be " << expectedType
+                  << ", but got " << resultType;
+
+      diag.attachNote(referencedMachine->getLoc())
+          << "original machine declared here";
+      return failure();
+    }
+  }
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// StateOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult StateOp::canonicalize(StateOp op, PatternRewriter &rewriter) {
+  bool hasAlwaysTakenTransition = false;
+  bool canonicalized = false;
+  SmallVector<TransitionOp, 4> transitionsToErase;
+  // Remove all transitions after the "always-taken" transition.
+  for (auto transition : op.transitions().getOps<TransitionOp>()) {
+    if (!hasAlwaysTakenTransition)
+      hasAlwaysTakenTransition = transition.isAlwaysTaken();
+    else {
+      transitionsToErase.push_back(transition);
+      canonicalized = true;
+    }
+  }
+  for (auto transition : transitionsToErase)
+    rewriter.eraseOp(transition);
+
+  return canonicalized ? success() : failure();
```

Comment:
instead of redundantly tracking "canonicalized", you simplify this to:

return failure(transitionsToErase.empty());